libs/data/bbhnet/data/glitch_sampler.py

```diff
@@ -1,90 +1,45 @@
-
-        self.hanford = torch.Tensor(hanford_glitches)
-        self.livingston = torch.Tensor(livingston_glitches)
-
-        self.deterministic = deterministic
-
-    def to(self, device: str) -> None:
-        self.hanford = self.hanford.to(device)
-        self.livingston = self.livingston.to(device)
-
-    def sample(self, N: int, size: int, offset: int = 0) -> np.ndarray:
-        """Sample glitches from each interferometer
-
-        If `self.deterministic` is `True`, this will grab the first
-        `N` glitches from each interferometer, with the center of
-        each kernel placed at the trigger time minus some specified
-        amount of offset.
-
-        If `self.deterministic` is `False`, this will sample _at most_
-        `N` kernels from each interferometer, with the _total_ glitches
-        sampled equal to `N`. The sampled glitches will be chosen at
-        random, and the kernel sampled from each glitch will be randomly
-        selected, with `offset` indicating the maximum distance the right
-        edge of the kernel can be from the trigger time, i.e. the default
-        value of 0 indicates that every kernel must contain the trigger.
-        """
-
-        if self.deterministic:
-            if N == -1:
-                N = len(self.hanford)
-
-            center = int(self.hanford.shape[-1] // 2)
-            left = int(center + offset - size // 2)
-            right = int(left + size)
-
-            hanford = self.hanford[:N, left:right]
-            livingston = self.livingston[:N, left:right]
-        else:
-            hanford = livingston = None
-            num_hanford = np.random.randint(N)
-            num_livingston = N - num_hanford
-
-            if num_hanford > 0:
-                hanford = sample_kernels(
-                    self.hanford, size, offset, num_hanford
+        super().__init__()
+        self.glitches = torch.nn.ParameterList()
+        for ifo in glitches.values():
+            param = self.add_parameter(ifo)
+            self.glitches.append(param)
+
+        self.prob = prob
+        self.max_offset = max_offset
+
+    def forward(self, X: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        if X.shape[1] < len(self.glitches):
+            raise ValueError(
+                "Can't insert glitches into tensor with {} channels "
+                "using glitches from {} ifos".format(
+                    X.shape[1], len(self.glitches)
                 )
-                hanford = torch.stack(hanford, axis=0)
-
-            if num_livingston > 0:
-                livingston = sample_kernels(
-                    self.livingston, size, offset, num_livingston
-                )
-                livingston = torch.stack(livingston, axis=0)
-
-        return hanford, livingston
+            )
+
+        masks = torch.rand(size=(len(self.glitches), len(X))) < self.prob
+        # masks = masks.to(X.device)
+        for i, ifo in enumerate(self.glitches):
+            mask = masks[i]

```

Comment:
So this mask determines which background elements are replaced with glitches

Reply:
yeah exactly. Marginally faster to just generate one mask up front (I think, maybe the slice ends up costing you but I'm fairly confident this is the faster way of doing it)